src/main/java/calculator/NumberExtractor.java

```diff
@@ -0,0 +1,45 @@
+package calculator;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class NumberExtractor {
+
+    private String input;
+    private String divider = "[,:]";
+
+    private final Matcher MATCHER;
+
+    public NumberExtractor(String input) {
+        this.input = input;
+        Pattern pattern = Pattern.compile("//(.)\n(.*)");
```

Comment:
Pattern은 비싼 객체인데, pattern을 아래와 같이 생성자 바깥에 정의해서 사용하는 건 어떨까요?
```suggestion
    private static final Pattern CUSTOM_SPLIT_REGEX = Pattern.compile("//(.)\n(.*)");
    this.MATCHER = CUSTOM_SPLIT_REGEX.matcher(input);
```

Reply:
인텔리제이 자동 완성에서 내부로 넣어도 된다는 추천을 아무 생각 없이 반영 했었네요 ㅜ
혹시 Pattern 객체가 비싼 객체라는 부분은 어느 부분을 보고 파악 하셨는지 알려주실 수 있을까요? 제가 아무 생각 없이 내장된 자바 객체를 구분 없이 자주 사용하는 경우가 많아서 이 부분을 고치기 위해 말씀주신 것처럼 막 만들어도 되는 객체, 비싸니까 자제해야 하는 객체를 구분할 수 있는 팁 알려주시면 감사하겠습니다!

Reply:
백기선님이 정리한 effective java 내용을 봤던 기억이 나서 리뷰를 남겼었어요 관련 링크도 같이 넣었어야 하는데 ㅠㅠ 아래 링크를 한번 참고해보시면 좋을것 같네요!

```java
String.matches가 가장 쉽게 정규 표현식에 매치가 되는지 확인하는 방법이긴 하지만 성능이 중요한 상황에서 반복적으로 사용하기에 적절하지 않다.

String.matches는 내부적으로 Pattern 객체를 만들어 쓰는데 그 객체를 만들려면 정규 표현식으로 유한 상태 기계로 컴파일 하는 과정이 필요하다. 즉 비싼 객체다.

성능을 개선하려면 Pattern 객체를 만들어 재사용하는 것이 좋다.

public class RomanNumber {

    private static final Pattern ROMAN = Pattern.compile("^(?=.)M*(C[MD]|D?C{0,3})(X[CL]|L?X{0,3})(I[XV]|V?I{0,3})$");

    static boolean isRomanNumeral(String s) {
        return ROMAN.matcher(s).matches();
    }

}
하지만 이 코드도 문제가 있는데, isRomanNumeral 메소드가 호출되지 않는다면, ROAM이라는 필요없이 만든셈이 된다. 게으른 초기화(lazily initializing)(아이템83)를 사용해서 최적화 할 수 있지만 추천하진 않는다. 보통 지연 초기화는 측정 가능한 성능 개선 없이 구현을 복잡하게 만든다.(아이템67)
```

https://github.com/keesun/study/blob/master/effective-java/item6.md

Reply:
상세한 답변 감사드립니다
첨부주신 자료랑 이펙티브 자바 서적 읽어보겠습니다 ㅎㅎ